pkg/azure/blob.go

```diff
@@ -0,0 +1,328 @@
+	if secret == nil {
+		return nil
+	}
+
+	var valid bool
+	if _, hasTenantID := secret.Data[tenantIDField]; hasTenantID {
+		if _, hasClientID := secret.Data[clientIDField]; hasClientID {
+			if _, hasClientSecret := secret.Data[clientSecretField]; hasClientSecret {
+				valid = true
+			}
+			if _, hasClientCertificate := secret.Data[clientCertificateField]; hasClientCertificate {
+				valid = true
+			}
+		}
+	}
+	if _, hasResourceID := secret.Data[resourceIDField]; hasResourceID {
+		valid = true
+	}
+	if _, hasClientID := secret.Data[clientIDField]; hasClientID {
+		valid = true
+	}
+	if _, hasAccountKey := secret.Data[accountKeyField]; hasAccountKey {
+		valid = true
+	}
+
+	if !valid {
+		return fmt.Errorf("invalid '%s' secret data: requires a '%s', '%s', or '%s' field, a combination of '%s', '%s' and '%s', or '%s', '%s' and '%s'",
+			secret.Name, resourceIDField, clientIDField, accountKeyField, tenantIDField, clientIDField, clientSecretField, tenantIDField, clientIDField, clientCertificateField)
+	}
+	return nil
+}
+
+// BucketExists returns if an object storage bucket with the provided name
+// exists, or returns a (client) error.
+func (c *BlobClient) BucketExists(ctx context.Context, bucketName string) (bool, error) {
+	container := c.ServiceClient.NewContainerClient(bucketName)
+	_, err := container.GetProperties(ctx, nil)
+	if err != nil {
+		var stgErr *azblob.StorageError
+		if errors.As(err, &stgErr) {
+			if stgErr.ErrorCode == azblob.StorageErrorCodeContainerNotFound {
+				return false, nil
+			}
+			err = stgErr
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// FGetObject gets the object from the provided object storage bucket, and
+// writes it to targetPath.
+// It returns the etag of the successfully fetched file, or any error.
+func (c *BlobClient) FGetObject(ctx context.Context, bucketName, objectName, localPath string) (string, error) {
+	container := c.ServiceClient.NewContainerClient(bucketName)
+	blob := container.NewBlobClient(objectName)
+
+	// Verify if destination already exists.
+	dirStatus, err := os.Stat(localPath)
+	if err == nil {
+		// If the destination exists and is a directory.
+		if dirStatus.IsDir() {
+			return "", ErrorDirectoryExists
+		}
+	}
+
+	// Proceed if file does not exist, return for all other errors.
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return "", err
+		}
+	}
+
+	// Extract top level directory.
+	objectDir, _ := filepath.Split(localPath)
+	if objectDir != "" {
+		// Create any missing top level directories.
+		if err := os.MkdirAll(objectDir, 0700); err != nil {
+			return "", err
+		}
+	}
+
+	// Download object.
+	res, err := blob.Download(ctx, nil)
+	if err != nil {
+		return "", err
+	}
+
+	// Prepare target file.
+	f, err := os.OpenFile(localPath, os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		return "", err
+	}
+
+	// Calculate hash during write.
+	// NOTE: not actively used at present, as MD5 is not consistently returned
+	// by API.
+	hash := md5.New()
```

Comment:
Azure's API support either md5 or crc64 for integrity checks - but I am not sure whether both can be simultaneously disabled. It might be worth validating them based on the values of `ContentCRC64` and `ContentMD5` in `BlobDownloadResult`.

Reply:
When I wrote the tests, I checked what values where returned for the most simple file PUT, and none of the fields seem to contain any value by default.